File: hilti/python/hilti/codegen/debug.py
# $Id$
#
# Code generator for debug instructions. 

import llvm.core

from hilti.core import *
from hilti import instructions
from codegen import codegen

import sys

@codegen.when(instructions.debug.Msg)
def _(self, i):
    if codegen.debugMode():
        self.llvmGenerateCCallByName("hlt::debug_printf", [i.op1(), i.op2(), i.op3()])
    
@codegen.when(instructions.debug.Assert)
def _(self, i):
    if codegen.debugMode():
        op1 = self.llvmOp(i.op1())
        block_true = self.llvmNewBlock("true")
        block_false = self.llvmNewBlock("false")

        self.builder().cbranch(op1, block_true, block_false)
        
        self.pushBuilder(block_false)
        arg = codegen.llvmOp(i.op2()) if i.op2() else string._makeLLVMString("")
        self.llvmRaiseExceptionByName("hlt_exception_assertion_error", i.location(), arg)
        self.popBuilder()

        self.pushBuilder(block_true)
        # Leave on stack. 
        
@codegen.when(instructions.debug.InternalError)
def _(self, i):
    arg = codegen.llvmOp(i.op1())
    self.llvmRaiseExceptionByName("hlt_exception_internal_error", i.location(), arg)

File: hilti/codegen/instructions/debug.cc

#include <hilti-intern.h>

#include "../stmt-builder.h"
#include "options.h"

using namespace hilti;
using namespace codegen;

void StatementBuilder::visit(statement::instruction::debug::Assert* i)
{
    if ( cg()->options().debug == 0 )
        return;

    auto op1 = cg()->llvmValue(i->op1());
    auto block_true = cg()->newBuilder("true");
    auto block_false = cg()->newBuilder("false");

    cg()->llvmCreateCondBr(op1, block_true, block_false);

    cg()->pushBuilder(block_false);
    auto arg = i->op2() ? cg()->llvmValue(i->op2()) : cg()->llvmString("");
    cg()->llvmRaiseException("Hilti::AssertionError", i->op1(), arg);
    cg()->popBuilder();

    cg()->pushBuilder(block_true);

    // Leave on stack.
}

void StatementBuilder::visit(statement::instruction::debug::InternalError* i)
{
    cg()->llvmRaiseException("Hilti::InternalError", i->location(), cg()->llvmValue(i->op1()));
}

void StatementBuilder::visit(statement::instruction::debug::Msg* i)
{
    CodeGen::expr_list args = { i->op1(), i->op2(), i->op3() };
    cg()->llvmCall("hlt::debug_printf", args);
}

void StatementBuilder::visit(statement::instruction::debug::PopIndent* i)
{
    cg()->llvmDebugPopIndent();
}

void StatementBuilder::visit(statement::instruction::debug::PushIndent* i)
{
    cg()->llvmDebugPushIndent();
}

